test/testaudiocapture.c

```diff
@@ -199,38 +122,91 @@ int main(int argc, char **argv)
-        SDL_Quit();
-        exit(1);
+        return -1;
     } else if (SDL_BindAudioStream(device, stream_in) == -1) {
         SDL_LogError(SDL_LOG_CATEGORY_APPLICATION, "Couldn't bind an audio stream for capture: %s!\n", SDL_GetError());
-        SDL_Quit();
-        exit(1);
+        return -1;
     }
 
     SDL_SetAudioStreamFormat(stream_in, NULL, &outspec);  /* make sure we output at the playback format. */
 
     SDL_Log("Ready! Hold down mouse or finger to record!\n");
 
-#ifdef __EMSCRIPTEN__
-    emscripten_set_main_loop(loop, 0, 1);
-#else
-    while (!done) {
-        loop();
-        if (!done) {
-            SDL_Delay(16);
+    return 0;
+}
+
+int SDL_AppEvent(const SDL_Event *event)
+{
+    if (event->type == SDL_EVENT_QUIT) {
+        return 1;  /* terminate as success. */
+    } else if (event->type == SDL_EVENT_KEY_DOWN) {
+        if (event->key.keysym.sym == SDLK_ESCAPE) {
+            return 1;  /* terminate as success. */
+        }
+    } else if (event->type == SDL_EVENT_MOUSE_BUTTON_DOWN) {
+        if (event->button.button == 1) {
+            SDL_PauseAudioDevice(SDL_GetAudioStreamDevice(stream_out));
+            SDL_FlushAudioStream(stream_out);  /* so no samples are held back for resampling purposes. */
+            SDL_ResumeAudioDevice(SDL_GetAudioStreamDevice(stream_in));
+        }
+    } else if (event->type == SDL_EVENT_MOUSE_BUTTON_UP) {
+        if (event->button.button == 1) {
+            SDL_PauseAudioDevice(SDL_GetAudioStreamDevice(stream_in));
+            SDL_FlushAudioStream(stream_in);  /* so no samples are held back for resampling purposes. */
+            SDL_ResumeAudioDevice(SDL_GetAudioStreamDevice(stream_out));
         }
     }
-#endif
+    return 0;  /* keep going. */
+}
 
-    SDLTest_CommonDestroyState(state);
+int SDL_AppIterate(void)
+{
+    if (!SDL_AudioDevicePaused(SDL_GetAudioStreamDevice(stream_in))) {
+        SDL_SetRenderDrawColor(renderer, 0, 255, 0, 255);
+    } else {
+        SDL_SetRenderDrawColor(renderer, 255, 0, 0, 255);
+    }
+    SDL_RenderClear(renderer);
+    SDL_RenderPresent(renderer);
 
-    return 0;
+    /* Feed any new data we captured to the output stream. It'll play when we unpause the device. */
+    while (SDL_GetAudioStreamAvailable(stream_in) > 0) {
+        Uint8 buf[1024];
+        const int br = SDL_GetAudioStreamData(stream_in, buf, sizeof(buf));
+        if (br < 0) {
+            SDL_LogError(SDL_LOG_CATEGORY_APPLICATION, "Failed to read from input audio stream: %s\n", SDL_GetError());
+            return -1;   /* quit the app, report failure. */
+        } else if (SDL_PutAudioStreamData(stream_out, buf, br) < 0) {
+            SDL_LogError(SDL_LOG_CATEGORY_APPLICATION, "Failed to write to output audio stream: %s\n", SDL_GetError());
+            return -1;   /* quit the app, report failure. */
+        }
+    }
+
+    SDL_Delay(10);  /* not a lot going on here, give up CPU time. */
```

Comment:
Do we need to do this anymore? What happens if the audio is opened with a smaller buffer length than our loop iteration time? Do we continually underflow?

Reply:
In this case, data will queue in the capture audio stream until the next iteration, and then dump it all into the playback stream, so even if this had extremely small internal buffers, the audio streams would make up the difference until the next iteration.

But I've moved this to use the hint that tells SDL that AppIterate doesn't need a lot of CPU and should be called less, in cd335a3290d7426e8dceaff52e18462acf8185bd